### PR TITLE
build: cbindgen (v2)

### PR DESCRIPTION
**Pull request by dbcfd**
Version 2 (from https://github.com/OISF/suricata/pull/3971)

Rust headers are now generated using cbindgen. If cbindgen is present, they can
be generated during dist, otherwise they will be available for builds.

Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

Describe changes:
- Rust headers are now created by cbindgen
- Cbindgen is run as a step under dist when cbindgen is available and a high enough version

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -1,5 +1,6 @@
     AC_INIT([suricata],[5.0.0-dev])
     m4_ifndef([AM_SILENT_RULES], [m4_define([AM_SILENT_RULES],[])])AM_SILENT_RULES([yes])
+    m4_pattern_forbid([AX_])
```
Comment (victorjulien) [inline configure.ac]: What is this for?

File: configure.ac [hunk 2]
```diff
@@ -2444,8 +2434,31 @@ fi
         echo "       for building the distribution"
         echo "   To install: cargo install cargo-vendor"
       fi
+      AC_CHECK_FILES([$srcdir/rust/gen], [have_rust_headers="yes"])
```
Comment (victorjulien) [inline configure.ac]: if we do have the headers already, doesn't that mean we can skip the block directly below?

File: rust/cbindgen.toml [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+#
+# default: "C++"
+language = "C"
+
+
+# Options for wrapping the contents of the header:
+
+# An optional string of text to output at the beginning of the generated file
+# default: doesn't emit anything
+header = """/* Copyright (C) 2017 Open Information Security Foundation
```
Comment (victorjulien) [inline rust/cbindgen.toml]: 2019

File: rust/cbindgen.toml [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+
+# A list of substitutions for converting cfg's to ifdefs. cfgs which aren't
+# defined here will just be discarded.
+#
+# e.g.
+# `#[cfg(target = "freebsd")] ...`
+# becomes
+# `#if defined(DEFINE_FREEBSD) ... #endif`
+[defines]
+"target_os = freebsd" = "DEFINE_FREEBSD"
```
Comment (victorjulien) [inline rust/cbindgen.toml]: Not sure I understand what is being done here with freebsd and serde?

File: rust/gen/rust-bindings.h [hunk 5]
```diff
@@ -0,0 +1,919 @@
+/* Copyright (C) 2017 Open Information Security Foundation
```
Comment (victorjulien) [inline rust/gen/rust-bindings.h]: should this be checked in at all? Seems more like something for the dist to contain a generated version of this from when the dist is built.

File: src/app-layer-dns-common.h [hunk 6]
```diff
@@ -121,8 +121,8 @@ enum {
 };
 
 /** Opaque Rust types. */
-typedef struct RSDNSState_ RSDNSState;
-typedef struct RSDNSTransaction_ RSDNSTransaction;
+//typedef struct RSDNSState_ RSDNSState;
```
Comment (victorjulien) [inline src/app-layer-dns-common.h]: Can be removed completely.

File: src/app-layer-parser.h [hunk 7]
```diff
@@ -94,11 +95,11 @@ typedef int (*AppLayerParserFPtr)(Flow *f, void *protocol_state,
         uint8_t *buf, uint32_t buf_len,
         void *local_storage, const uint8_t flags);
 
-typedef struct AppLayerGetTxIterTuple {
-    void *tx_ptr;
-    uint64_t tx_id;
-    bool has_next;
-} AppLayerGetTxIterTuple;
+//typedef struct AppLayerGetTxIterTuple {
```
Comment (victorjulien) [inline src/app-layer-parser.h]: Please remove

File: src/output-json-dhcp.c [hunk 8]
```diff
@@ -78,7 +78,7 @@ static int JsonDHCPLogger(ThreadVars *tv, void *thread_data,
         return TM_ECODE_FAILED;
     }
 
-    json_t *dhcp_js = rs_dhcp_logger_log(ctx->rs_logger, tx);
+    json_t *dhcp_js = (json_t *)rs_dhcp_logger_log(ctx->rs_logger, tx);
```
Comment (victorjulien) [inline src/output-json-dhcp.c]: can we avoid these casts?

File: src/rust.h [hunk 9]
```diff
@@ -51,9 +58,9 @@ struct _Store;
 typedef struct _Store Store;
 
 /** Opaque Rust types. */
-typedef struct NFState_ NFSState;
-typedef struct NFSTransaction_ NFSTransaction;
-typedef struct SMBState_ SMBState;
-typedef struct SMBTransaction_ SMBTransaction;
+//typedef struct NFState_ NFSState;
```
Comment (victorjulien) [inline src/rust.h]: Please remove.

File: src/output-json-dhcp.c (hunk 8 above)
Comment (jasonish) [reply src/output-json-dhcp.c]: Looks like we could do something like this in `cbindgen.toml`:
```
[export]
exclude = ["JsonT"]

[export.rename]
"JsonT" = "json_t"
```

Comment (dbcfd) [reply src/output-json-dhcp.c]: that worked

**Comment by dbcfd** [comment]
Closing for https://github.com/OISF/suricata/pull/4244

**Pull request closed without merging**